core/chapters/c12_dictionaries.py

```diff
@@ -0,0 +1,510 @@
+
+        tests = [
+            ((['apple', 'box', 'cat'],  {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100}), 107)
+        ]
+
+        hints = "TODO"
+
+    class shopping_cart4(VerbatimStep):
+        """
+Perfect! You publish your website and start dreaming about how rich you're going to be.
+
+But soon you get a complaint from a customer who wants to buy 5 million dogs...and 2 boxes to put them in.
+
+Your website allows buying the same items several times, e.g. `total_cost(['box', 'box'], {...})` works,
+but they have to add each item one at a time, and for some reason this customer doesn't want to click
+'Add to Cart' 5 million times. People are so lazy!
+
+Here's the new code for you to fix:
+
+    __copyable__
+    __program_indented__
+        """
+
+        def program(self):
+            def total_cost(cart, quantities, prices):
+                result = 0
+                for item in cart:
+                    price = ...
+                    quantity = ...
+                    result += price * quantity
+                return result
+
+            assert_equal(
+                total_cost(
+                    ['dog', 'box'],
+                    {'dog': 5000000, 'box': 2},
+                    {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100},
+                ),
+                500000010,
+            )
+
+
+    class shopping_cart5(VerbatimStep):
+        """
+We've added another parameter called `quantities` to `total_cost`.
+Now `cart` is still a list of strings, but it doesn't have any duplicates.
+`quantities` is a dictionary where the keys are the items in `cart` and the corresponding values are the quantity
+of that item that the customer wants to buy.
+
+
+Not bad! But you may have noticed that it looks a bit awkward. Why do we have to specify `'dog'` and `'box'` in both the `cart` and the `quantities`?
+
+On the next page we'll look at how to loop directly over the keys of a dictionary,
+so we can get rid of the `cart` argument.
+
+
+But first, let's practice what we've learned a bit more.
+
+[Earlier in the course](#IntroducingElif) we looked at converting one strand of DNA
+into a new strand with matching nucleotides:
+
+    __copyable__
+    __program_indented__
+        """
+
+        def program(self):
+            def substitute(string):
+                result = ''
+                for char in string:
+                    if char == 'A':
+                        char = 'T'
+                    elif char == 'T':
+                        char = 'A'
+                    elif char == 'G':
+                        char = 'C'
+                    elif char == 'C':
+                        char = 'G'
+                    result += char
+                return result
+
+            original = 'AGTAGCGTCCTTAGTTACAGGATGGCTTAT'
+            expected = 'TCATCGCAGGAATCAATGTCCTACCGAATA'
+            assert_equal(substitute(original), expected)
+
+
+    ##TODO: Is this supposed to be an exercise or a verbatim step?
```

Comment:
A VerbatimStep means "run this exact code", usually shown to the user. `shopping_cart4` and `dna_part2` need to be exercise steps because the user has to figure out what to write to fill in the `...`, and there are multiple possible correct answers.